cmd/samFilter/samFilter.go

```diff
@@ -0,0 +1,296 @@
+	if int(a.Flag) != filter {
+		return false
+	}
+	return true
+}
+
+func filterByRegion(a sam.Sam, filter bed.Bed) bool {
+	if bed.Overlap(filter, convert.SamToBed(a)) {
+		return true
+	} else {
+		return false
+	}
+}
+
+func collapseUMI(inMap map[string]sam.Sam, inRead sam.Sam) map[string]sam.Sam {
+	umiMap := inMap
+	sc := sam.ToSingleCellAlignment(inRead)
+	umiBxString := fmt.Sprintf("%s%s", dna.BasesToString(sc.Bx), dna.BasesToString(sc.Umi))
+	_, found := umiMap[umiBxString]
+	if !found { //if the UMI isn't in the map, add it.
+		umiMap[umiBxString] = inRead
+	} else {
+		if umiMap[umiBxString].MapQ > inRead.MapQ {
+			//if the UMI is in the map and the UMI in the map has a better alignment score -- do nothing.
+		} else if umiMap[umiBxString].MapQ == inRead.MapQ {
+			if umiMap[umiBxString].Pos >= inRead.Pos { //if the UMIs are tied on alignment score, prioritize the one with a higher start position since GWAS STARR-seq construct barcodes are on the end of constructs,
+				//higher chance that the read contains the construct barcode.
+				//do nothing
+			} else if umiMap[umiBxString].Pos < inRead.Pos { //if the UMI in the map has a lower position than the query, replace it.
+				umiMap[umiBxString] = inRead
+			}
+		} else if umiMap[umiBxString].MapQ < inRead.MapQ { //if the UMI in the map has a lower alignment score than the query, replace it.
+			umiMap[umiBxString] = inRead
+		}
+	}
+	return umiMap
+}
+
+func appendUmiCbc(a sam.Sam) (sam.Sam, bool) {
+	var found bool = false
+	var newReadName string
+
+	cbc, cbcFound, _ := sam.QueryTag(a, "CB")
+	umi, umiFound, _ := sam.QueryTag(a, "UB")
+	cbcString := fmt.Sprintf("%s", cbc)
+	w := strings.Split(cbcString, "-")
+
+	if cbcFound && umiFound {
+		found = true
+		newReadName = fmt.Sprintf("%s_UMI:%s_BX:%s", a.QName, umi, w[0]) //compatible with scCount.go
+		a.QName = newReadName
+	}
+	return a, found
+}
+
+func runFilter(s Settings) {
+	var passQual, passLen, passCigar, foundUMI, passFlag, passLocation bool
+	var umiMap = make(map[string]sam.Sam)
+	var words, p []string
+	var chrom string
+	var start, end int
+	var bedRegion bed.Bed
+	var outSlice []sam.Sam
+	var err error
+	var bw *sam.BamWriter
+
+	inChan, header := sam.GoReadToChan(s.InFile)
+	out := fileio.EasyCreate(s.OutFile)
+	if s.OutBam {
+		bw = sam.NewBamWriter(out, header)
+	} else if s.OutFile != "stdout" && s.NoHeader == false {
+		sam.WriteHeaderToFileHandle(out, header)
+	}
+	if s.FilterByRegion != "" {
+		words = strings.Split(s.FilterByRegion, ":")
+		chrom = words[0]
+		if len(words) == 1 {
+			bedRegion = bed.Bed{Chrom: chrom, ChromStart: 0, ChromEnd: numbers.MaxInt}
+		} else {
+			p = strings.Split(words[1], "-")
+			start = common.StringToInt(p[0])
+			end = common.StringToInt(p[1])
+			bedRegion = bed.Bed{Chrom: chrom, ChromStart: start, ChromEnd: end}
+		}
+	}
+	for i := range inChan {
+		if s.AlignQualFilter > 0 {
+			passQual = filterByQuality(i, s.AlignQualFilter)
```

Comment:
Unless you are going to use the `passN` variables later you could compress a lot of this code by doing `if !filterFunction(inputs) { continue }` instead of saving to a variable and testing the variable.